### PR TITLE
JDK-8313781: Add regression tests for large page logging and user-facing error messages

### PR DESCRIPTION
**Pull request by tstuefe**
In preparation for the UseSHM removal (https://bugs.openjdk.org/browse/JDK-8310841) I'd like more regression tests. 

The new test basically checks the behavior of the JVM when confronted with different variations of:
- OS capabilities
- JVM settings.

For example: If the JVM is started with -XX:+UseLargePages, but on the OS side there are no huge pages configured, or the kernel does not even support hugepages, the JVM should disable large pages and issue a warning.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313781](https://bugs.openjdk.org/browse/JDK-8313781): Add regression tests for large page logging and user-facing error messages (**Bug** - P4)

### Reviewers
 * [Stefan Johansson](https://openjdk.org/census#sjohanss) (@kstefanj - **Reviewer**) ⚠️ Review applies to [8cc04b03](https://git.openjdk.org/jdk/pull/15159/files/8cc04b03952fa43b6096d465d205ee2854a30676)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15159/head:pull/15159` \
`$ git checkout pull/15159`

Update a local copy of the PR: \
`$ git checkout pull/15159` \
`$ git pull https://git.openjdk.org/jdk.git pull/15159/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15159`

View PR using the GUI difftool: \
`$ git pr show -t 15159`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15159.diff">https://git.openjdk.org/jdk/pull/15159.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15159#issuecomment-1666465733)

### CODE REVIEW

**Comment by tstuefe** [comment]
Any takers?

File: test/hotspot/jtreg/runtime/os/TestHugePageDecisionsAtVMStartup.java
Comment (kstefanj) [inline test/hotspot/jtreg/runtime/os/TestHugePageDecisionsAtVMStartup.java:119]: I would prefer skipping the empty lines here and below.
```suggestion
        } else if (useLP && !useTHP &&
```

**Review by kstefanj — APPROVED** [review]
Looks good, just one minor thing. 

I saw the comment about `LargePageSizeInBytes` and thought that it should be required away but realized that `ProcessTools.createJavaProcessBuilder()` doesn't add the passed in options.

Also did some manual runs on an aarch64 system where the default 512m pages are not configured but the 2m pages are. This seem to generate output which will be accepted by the test.

**Comment by tstuefe** [comment]
Thanks @kstefanj!

Since we just changed the error warning texts (https://github.com/openjdk/jdk/pull/15158), I needed to change the error text somewhat.

I then retested on x64 with the following configurations:

- No Hugepages configured (but supported by kernel) and:
  - THP "never"
  - THP "always"
  - THP "advise"

- THP "never" and 2M default hugepage size and:
  - nonzero Hugepages enabled (2M)
  - nonzero Hugepages enabled (1G)
  - nonzero Hugepages enabled (2M, 1G)

- THP "never" and 1G default hugepage size (via kernel parameter) and nonzero 1G Hugepages
 ( this uncovered a small test bug where "buildSizeString()" had been wrong)

- Hugepages completely absent on a kernel built without hugepages (aarch64 SBC)
	(this uncovered another test bug where HugePageConfiguration would fall back to THP mode "unknown" which then confused the test. I removed that enum value.)

All issues have been fixed.

File: test/hotspot/jtreg/runtime/os/TestHugePageDecisionsAtVMStartup.java
Comment (kstefanj) [inline test/hotspot/jtreg/runtime/os/TestHugePageDecisionsAtVMStartup.java:122]:
```suggestion
        } else if (useLP && useTHP && !configuration.supportsTHP()) {
```

File: test/hotspot/jtreg/runtime/os/TestHugePageDecisionsAtVMStartup.java
Comment (kstefanj) [inline test/hotspot/jtreg/runtime/os/TestHugePageDecisionsAtVMStartup.java:126]:
```suggestion
        } else if (useLP && !useTHP &&
```

File: test/hotspot/jtreg/runtime/os/TestHugePageDecisionsAtVMStartup.java
Comment (kstefanj) [inline test/hotspot/jtreg/runtime/os/TestHugePageDecisionsAtVMStartup.java:133]:
```suggestion
        } else if (useLP && useTHP && configuration.supportsTHP()) {
```

**Comment by tstuefe** [comment]
Thanks @kstefanj !

Took over your suggestions, they make the code look cleaner.

**Comment by tstuefe** [comment]
Friendly ping

**Comment by tstuefe** [comment]
Friendly ping. Only need a second review.

**Review by dholmes-ora — APPROVED** [review]
Seems okay. Thanks

**Comment by tstuefe** [comment]
Thank you, David!

/integrate

**Pull request closed without merging**
